packages/pricing/src/repositories/pricing.ts

```diff
@@ -0,0 +1,112 @@
+
+    // Quantity is used to scope money amounts based on min_quantity and max_quantity.
+    // We should potentially think of reserved words in pricingContext that can't be used in rules
+    // or have a separate pricing options that accept things like quantity, price_list_id and other
+    // pricing module features
+    const quantity = context.quantity
+    delete context.quantity
+
+    // Currency code here is a required param.
+    const currencyCode = context.currency_code
+    delete context.currency_code
+
+    if (!currencyCode) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `currency_code is a required input in the pricing context`
+      )
+    }
+
+    // Gets all the price set money amounts where rules match for each of the contexts
+    // that the price set is configured for
+    const psmaSubQueryKnex = knex({
+      psma: "price_set_money_amount",
+    })
+      .select({
+        id: "psma.id",
+        price_set_id: "psma.price_set_id",
+        money_amount_id: "psma.money_amount_id",
+        number_rules: "psma.number_rules",
+      })
+      .leftJoin("price_set_money_amount as psma1", "psma1.id", "psma.id")
+      .leftJoin("price_rule as pr", "pr.price_set_money_amount_id", "psma.id")
+      .leftJoin("rule_type as rt", "rt.id", "pr.rule_type_id")
+      .orderBy("number_rules", "desc")
+      .orWhere("psma1.number_rules", "=", 0)
+      .groupBy("psma.id")
+      .having(knex.raw("count(DISTINCT rt.rule_attribute) = psma.number_rules"))
+
+    for (const [key, value] of Object.entries(context)) {
+      psmaSubQueryKnex.orWhere({
+        "rt.rule_attribute": key,
+        "pr.value": value,
+      })
+    }
+
+    const priceSetQueryKnex = knex({
+      ps: "price_set",
+    })
+      .select({
+        id: "ps.id",
+        amount: "ma.amount",
+        min_quantity: "ma.min_quantity",
+        max_quantity: "ma.max_quantity",
+        currency_code: "ma.currency_code",
+        default_priority: "rt.default_priority",
+        number_rules: "psma.number_rules",
+      })
+      .join(psmaSubQueryKnex.as("psma"), "psma.price_set_id", "ps.id")
+      .join("money_amount as ma", "ma.id", "psma.money_amount_id")
+      .leftJoin("price_rule as pr", "pr.price_set_money_amount_id", "psma.id")
+      .leftJoin("rule_type as rt", "rt.id", "pr.rule_type_id")
+      .whereIn("ps.id", pricingFilters.id)
+      .andWhere("ma.currency_code", "=", currencyCode)
+      .orderBy([
+        { column: "number_rules", order: "desc" },
+        { column: "default_priority", order: "desc" },
+      ])
+
+    if (quantity) {
+      priceSetQueryKnex.where("ma.min_quantity", "<=", quantity)
+      priceSetQueryKnex.andWhere("ma.max_quantity", ">=", quantity)
+    }
+
+    const isContextPresent = Object.entries(context).length || !!currencyCode
+    // Only if the context is present do we need to query the database.
+    const queryBuilderResults = isContextPresent ? await priceSetQueryKnex : []
```

Comment:
suggestion: don't we know if the context is present at the begining? in that case Iwould return prematurally at the begining